[DOCS] Add NatSpec documentation

fulldecent opened the pull request:
Fixes #6121

stackenbotten commented:
There was an error when running `test_check_style` for commit `ece0c687aab2b9133c743f280e4d58e9897478df`:
```
Error: Trailing whitespace found:
 docs/natspec-format.rst:77:.. code:: 
 docs/natspec-format.rst:79: /// @notice Send `(valueInmGAV / 1000).fixed(0,3)` GAV from the account of 
 docs/natspec-format.rst:116:.. code:: 
 docs/natspec-format.rst:122: /// @notice Send `(valueInmGAV / 1000).fixed(0,3)` GAV from the account of 
 docs/natspec-format.rst:152:.. code:: 
 docs/natspec-format.rst:183:.. code:: 

```
Please check that your changes are working as intended.

fulldecent committed:
Add NatSpec documentation.

chriseth commented on docs/metadata.rst:
```diff
@@ -141,7 +141,7 @@ Furthermore, the wallet can use the NatSpec user documentation to display a conf
 authorization for the transaction signature.
 
-Additional information about Ethereum Natural Specification (NatSpec) can be found `here <https://github.com/ethereum/wiki/wiki/Ethereum-Natural-Specification-Format>`_.
+Additional information about Ethereum Natural Specification (NatSpec) can be found `here <https://solidity.readthedocs.io/en/latest/natspec-format.html>`_.
```
Please use proper internal links

chriseth commented on docs/natspec-format.rst:
```diff
@@ -0,0 +1,202 @@
+Solidity contracts can have a special form of comments that form the
```
I think this file needs a heading.

chriseth commented on docs/natspec-format.rst:
```diff
@@ -0,0 +1,202 @@
+   documentation <https://vyper.readthedocs.io/en/latest/structure-of-a-contract.html#natspec-metadata>`__.
+
+This example shows a contract and a function using all available tags.
+Note: NatSpec currently does NOT apply to variables (see
```
Are you talking about public state variables or variables in general?

chriseth commented on docs/natspec-format.rst:
```diff
@@ -0,0 +1,202 @@
+Solidity contracts can have a special form of comments that form the
+basis of the Ethereum Natural Specification Format.
```
This sounds like people already know what the Ethereum Natual Specification Format is. I think we should introduce the term here instead and explain what it is meant to be used for.

chriseth commented on docs/natspec-format.rst:
```diff
@@ -0,0 +1,202 @@
+   /// @author Warned Bros
+   /// @notice You can use this contract for only the most basic simulation
+   /// @dev All function calls are currently implement without side effects
+   contract BugsBunny {
```
I'm not sure we should use trademarks (I assume) in the documentation without the need for them.

chriseth commented on docs/natspec-format.rst:
```diff
@@ -0,0 +1,202 @@
+       }
+   }
+
+.. _header-n12:
```
Please use a descriptive name.

chriseth commented on docs/natspec-format.rst:
```diff
@@ -0,0 +1,202 @@
+Solidity contracts can have a special form of comments that form the
+basis of the Ethereum Natural Specification Format.
+
+.. _header-n3:
```
Please use a descriptive name.

chriseth commented on docs/natspec-format.rst:
```diff
@@ -0,0 +1,202 @@
+then use multiple ``@return`` statements in the same format as the
+``@param`` statements.
+
+.. _header-n44:
```
Please use a descriptive name.

chriseth commented on docs/natspec-format.rst:
```diff
@@ -0,0 +1,202 @@
+
+.. code::
+
+   /// @notice Send `(valueInmGAV / 1000).fixed(0,3)` GAV from the account of
```
It would be great if this could refer to the example above and not some weird coin nobody has heard of before ;)

chriseth replied:
In addition to the example, things like `.fixed(...)` have never worked.

chriseth commented on docs/natspec-format.rst:
```diff
@@ -0,0 +1,202 @@
+   ...
+
+If a user (address 0x2334) attempts to call this function with a ``to``
+address of ``0x0`` and ``valueInmGAV`` of 4,135 then this will render to
```
What is the meaning of the `,` in `4,134`? Any reason you use `,` here and `.` below?

chriseth commented on docs/natspec-format.rst:
```diff
@@ -0,0 +1,202 @@
+   Send 4.135 GAV from the account of 0.2334 to an account accessible
+   only by 0x0
+
+Use any Javascript/Paperscript expression encapsulated in backticks as
```
I stil don't know what Paperscript was meant to refer to...

chriseth commented on docs/natspec-format.rst:
```diff
@@ -0,0 +1,202 @@
+Dynamic expressions
+-------------------
+
+In function documentation, you may use dynamic expressions for all tags.
```
I think we should also mention the new effort of a more restrictive feature set in these expressions.

chriseth commented:
Sorry about the many comments. I think this is very useful, especially also mentioning how to use it in Vyper. I just realized rather late that you probably just copied the Wiki article, did you?

chriseth commented:
You can continue cleaning up the WIki article, but if you want, we can also take over. In any case, please rebase instead of merging.

ChrisChinchilla commented:
I think @fulldecent also created that Wiki page, as it was missing last week, and I was in the process of getting access to bring it back, to then find today that @fulldecent had recreated/reproduced it. I'm not really sure of what supposed to be in that wiki verses the Solidity docs, but any changes we do end up making to the content in this PR should also find their way to https://github.com/ethereum/wiki/wiki/Ethereum-Natural-Specification-Format or, we remove the page in that wiki and instead link to the page we create here to reduce duplication.

fulldecent commented:
Rebased, many updates made to improve RST formatting and content.

Added opinionated statement:

> It is recommended that Solidity contracts are fully annotated using NatSpec for all public interfaces (everything in the ABI).

Again, this is surely a best practice, but it would be better if this is said directly in the Solidity documentation.

Yes this was copy-pasted from the Wiki. Yes, I am a contributor to that wiki.

My main motivation is that the wiki is prone to vandalism and is not a place to publish standards. Solidity implements NatSpec and has an interest to see that it is documented properly.

I am hoping to get this PR up. And then continue to make improvements. Then I also want to propose a change to the behavior of solc related to NatSpec. But I can't do that until I can actually cite a specification for NatSpec!

ChrisChinchilla commented on docs/metadata.rst:
```diff
@@ -154,7 +154,7 @@ Furthermore, the wallet can use the NatSpec user documentation to display a conf
 authorization for the transaction signature.
 
-Additional information about Ethereum Natural Specification (NatSpec) can be found `here <https://github.com/ethereum/wiki/wiki/Ethereum-Natural-Specification-Format>`_.
+See additional information about the :doc:`Ethereum Natural Specification (NatSpec) format <natspec-format>`.
```
```suggestion
For additional information, read :doc:`Ethereum Natural Specification (NatSpec) format <natspec-format>`.
```

ChrisChinchilla commented:
@fulldecent Syntax changes generally look a lot better to me in the diff view, but I may do a little final polish, checking and testing locally, but also need some further input from @chriseth on technical side, wether we include the doc here etc etc.

fulldecent commented:
@ChrisChinchilla Thank you. Also you are set up so you can push into my branch as well.

stackenbotten commented:
There was an error when running `test_check_style` for commit `dd466f6142d73b5adebd2ddd3f83a65e1ae0fe29`:
```
Error: Trailing whitespace found:
 docs/style-guide.rst:1128: 

```
Please check that your changes are working as intended.

fulldecent commented:
Thank you. I believe all comments are handled at this point.

I have added my agenda for reforms as follow on items (i.e. after the instant issue/PR is closed) at https://github.com/ethereum/solidity/issues/6121

If/when this PR is merged then I will work on motivating/starting implementation for each of those items.

chriseth commented on docs/natspec-format.rst:
```diff
@@ -0,0 +1,216 @@
+
+Solidity contracts can use a special form of comments to provide rich
+documentation for functions, return variables and more. This special form is
+named the Ethereum Natural Specification Format (NatSpec).
```
```suggestion
named the Ethereum Natural Language Specification Format (NatSpec).
```

chriseth replied:
I think we should already say here that it follows JavaDoc/DoxyGen style.

chriseth commented on docs/natspec-format.rst:
```diff
@@ -0,0 +1,216 @@
+   comments. See `Vyper
+   documentation <https://vyper.readthedocs.io/en/latest/structure-of-a-contract.html#natspec-metadata>`__.
+
+This example shows a contract and a function using all available tags.
```
```suggestion
The following example shows a contract and a function using all available tags.
```
(it is not that close to the sentence)

chriseth commented on docs/natspec-format.rst:
```diff
@@ -0,0 +1,216 @@
+Note: NatSpec currently does NOT apply to public state variables (see
+`solidity#3418 <https://github.com/ethereum/solidity/issues/3418>`__),
+even if they are declared public and therefore do affect ABI. Note:
+NatSpec currently only interprets tags functions if they are external or
```
I think a word is missing here: "interprets tags functions"

fulldecent replied:
Updated in c476471

chriseth commented on docs/natspec-format.rst:
```diff
@@ -0,0 +1,216 @@
+even if they are declared public and therefore do affect ABI. Note:
+NatSpec currently only interprets tags functions if they are external or
+public. You are welcome to use similar comments for your internal and
+private functions, but those will not be parsed.
```
At this point, we do not yet know what it means for natspec comments to be parsed, do we?

fulldecent replied:
Updated in c476471

chriseth commented on docs/natspec-format.rst:
```diff
@@ -0,0 +1,216 @@
+====
+
+All tags are optional. The following table explains the purpose of each
+NatSpec tag and where it may be used.
```
If no tag is given, `@notice` is assumed.

fulldecent replied:
Updated in 8233930

chriseth commented on docs/natspec-format.rst:
```diff
@@ -0,0 +1,216 @@
+         "details": "The Alexandr N. Tetearing algorithm could increase precision"
+       }
+     },
+     "invariants": [
```
Is such a concept present somewhere?

fulldecent replied:
Yes, this is a real thing. https://en.wikipedia.org/wiki/Dendrochronology

I went out to learn some biology rather than show some useless contrived example.

chriseth replied:
No, I mean the invariant concept - is that implemented in one of the languages that support natspec?

chriseth replied:
About the example: Yes, I really like it, too!

fulldecent replied:
Thanks! Ah, no the invariant is cruft. I need to rerun the program to get the correct output. Patch coming...

fulldecent replied:
Updated in 5c0eabe

stackenbotten commented:
There was an error when running `test_check_style` for commit `5c0eabe3bd25ac23d959ea1609cccd322902b1e2`:
```
Error: Trailing whitespace found:
 docs/natspec-format.rst:137: "methods" : 
 docs/natspec-format.rst:139: "age(uint256)" : 
 docs/natspec-format.rst:165: "methods" : 
 docs/natspec-format.rst:167: "age(uint256)" : 
 docs/natspec-format.rst:171: "params" : 

```
Please check that your changes are working as intended.

chriseth commented on docs/natspec-format.rst:
```diff
@@ -0,0 +1,180 @@
+documentation for functions, return variables and more. This special form is
+named the Ethereum Natural Language Specification Format (NatSpec).
+
+This documentation is segmented into developer-focused messages and end-user-
```
Will the newline create a space in the rendered document?

fulldecent replied:
This is currently undefined behavior as far as how I have written the documentation. But the actual implementation is Solidity is that newlines in the docblock are concatenated with the newline deleted. A full demonstration follows.

Our options at this point are:

1. Leave the issue of whitespace as undefined behavior for #6122 <kbd>RECOMMENDED</kbd>
2. Add whitespace handling to the specification

Please advise.

# Test case

```solidity
pragma solidity ^0.5.6;

/// @title A simulator for trees
/// extended title
/// @author Larry A. Gardner
///  extended author
/// @notice You can use this contract for only the most basic simulation
///         extended notice
/// @dev All function calls are currently implemented without side effects
/// @dev Repeated dev
contract Tree {
    /// @author Mary A. Botanist
    /// @notice Calculate tree age in years, rounded up, for live trees
    /// @dev The Alexandr N. Tetearing algorithm could increase precision
    /// @param rings The number of rings from dendrochronological sample
    /// @return age in years, rounded up for partial years
    function age(uint256 rings) external pure returns (uint256) {
        return rings + 1;
    }
}
```

Rendered output:

```json
{
  "author" : "Larry A. Gardner  extended author",
  "details" : "All function calls are currently implemented without side effects Repeated dev",
  "methods" : 
  {
    "age(uint256)" : 
    {
      "author" : "Mary A. Botanist",
      "details" : "The Alexandr N. Tetearing algorithm could increase precision",
      "params" : 
      {
        "rings" : "The number of rings from dendrochronological sample"
      },
      "return" : "age in years, rounded up for partial years"
    }
  },
  "title" : "A simulator for trees extended title"
}
```

and

```
{
  "methods" : 
  {
    "age(uint256)" : 
    {
      "notice" : "Calculate tree age in years, rounded up, for live trees"
    }
  },
  "notice" : "You can use this contract for only the most basic simulation         extended notice"
}
```

chriseth replied:
I think this was a misunderstanding. I was talking about the rst rendering, not about natspec.

chriseth commented on docs/natspec-format.rst:
```diff
@@ -0,0 +1,180 @@
+
+NatSpec includes the formatting for comments that the smart contract author will
+use, and which are understood by the Solidity compiler. Also detailed below is
+output of the Solidity compiler, which extracts these comments into a machine-
```
Same here.

chriseth commented on docs/natspec-format.rst:
```diff
@@ -0,0 +1,180 @@
+The following example shows a contract and a function using all available tags.
+Note: NatSpec currently does NOT apply to public state variables (see
+`solidity#3418 <https://github.com/ethereum/solidity/issues/3418>`__),
+even if they are declared public and therefore do affect ABI. Note:
```
```suggestion
even if they are declared public and therefore do affect the ABI. Note:
```

chriseth commented on docs/natspec-format.rst:
```diff
@@ -0,0 +1,180 @@
+Tags
+====
+
+All tags are optional. The following table explains the purpose of each
```
Do you know that there are people who re-formulate their sentences until all lines render with exactly the same width? It is the case for this paragraph, so I checked the others and was delighted to see that this just seems to be a coincidence :)

fulldecent replied:
Who can say that it was a coincidence!?

chriseth commented on docs/natspec-format.rst:
```diff
@@ -0,0 +1,180 @@
+``@return`` Documents the return type of a contract's function                              function
+=========== =============================================================================== =============================
+
+If your function returns a tuple, like ``(int quotient, int remainder)``
```
We don't really call these tuples in other parts of the documentation, so perhaps rather talk about "returns multiple variables like ``(int...`.

fulldecent replied:
Updated in 6667d17633ab6dc13a4e504e21292fa09163ea9c

chriseth commented on docs/natspec-format.rst:
```diff
@@ -0,0 +1,180 @@
+Dynamic expressions
+-------------------
+
+In function documentation, you may use dynamic expressions in any tags.
```
I think you should provide an example here, without context, "dynamic expressions" can mean anything. Perhaps also rather call them "evaluated expressions" or "expressions to be evaluated" or "templates"?

fulldecent replied:
Example and explanation provided in 9cfd68763bc8248140d2eba73b4e58bd40ae6082

chriseth commented on docs/natspec-format.rst:
```diff
@@ -0,0 +1,180 @@
+====================
+
+When parsed by the compiler, documentation such as the one from the
+above example will produce 2 different JSON files. One is meant to be
```
```suggestion
above example will produce two different JSON files. One is meant to be
```

chriseth replied:
Please change this, I think it hinders reading flow.

chriseth commented on docs/natspec-format.rst:
```diff
@@ -0,0 +1,180 @@
+consumed by the end user as a notice when a function is executed and the
+other to be used by the developer.
+
+If the above contract is saved as ex1.sol then you can generate the
```
```suggestion
If the above contract is saved as ``ex1.sol`` then you can generate the
```

chriseth reviewed (changes requested):
Wonderful, thanks a lot for your continuing effort! Just some small change requests left.

fulldecent commented:
@chriseth Thank you for the feedback. I believe all issues are addressed now.
